### PR TITLE
feat(cctp): arbitrary L1 -> L2 finalizations

### PR DESCRIPTION
**Pull request by james-a-morris**
This PR brings the L1 -> L2 CCTP finalizations in-line with the Linea finalizer in that we should be able to choose an arbitrary list of addresses to finalize originating from mainnet.

Hashes to confirm this routing:
1. https://basescan.org/tx/0xd7616ab716fbaff5c82d033106ffc3e4ab8dd17c2c65942ed42e0ee97578b75e
2. https://arbiscan.io/tx/0x82a4db0e85c7c4bbe1d55d1ad9690ba1b48db756ef4af6fe625592536573a560
3. https://optimistic.ethercan.io/tx/0xe48117748ffbb5b70ab609640c93adf61941e436acef72e107402244a54e9a5f

### CODE REVIEW

File: src/finalizer/utils/cctp/l1ToL2.ts [hunk 1]
```diff
@@ -1,58 +1,95 @@
...
+): Promise<TransactionReceipt[]> {
+  const provider = getCachedProvider(currentChainId);
+  const tokenMessengerContract = new Contract(
+    CONTRACT_ADDRESSES[currentChainId].cctpTokenMessenger.address,
+    CONTRACT_ADDRESSES[currentChainId].cctpTokenMessenger.abi,
+    provider
+  );
+  const eventFilter = tokenMessengerContract.filters.DepositForBurn(
+    undefined,
+    undefined, // This should be the UDSC address, but we are currently blocked
```
Comment (james-a-morris) [inline src/finalizer/utils/cctp/l1ToL2.ts:68]: Blocked on https://github.com/across-protocol/constants-v2/pull/28 to specify USDC the address.

We could technically leave this as-is to allow our monitor addresses to pass more than USDC via CCTP.

**Review by james-a-morris — COMMENTED** [review]
Comments on Code

File: src/finalizer/utils/cctp/l1ToL2.ts (hunk 1 above)
Comment (nicholaspai) [reply src/finalizer/utils/cctp/l1ToL2.ts:68]: would we ever monitor any other address besides USDC on CCTP...?

File: src/finalizer/index.ts [hunk 2]
```diff
@@ -63,15 +63,14 @@ const chainFinalizers: { [chainId: number]: ChainFinalizer } = {
  */
 const chainFinalizerOverrides: { [chainId: number]: ChainFinalizer[] } = {
   // Mainnets
-  1: [lineaL1ToL2Finalizer],
   10: [opStackFinalizer, cctpL1toL2Finalizer, cctpL2toL1Finalizer],
   137: [polygonFinalizer, cctpL1toL2Finalizer, cctpL2toL1Finalizer],
   8453: [opStackFinalizer, cctpL1toL2Finalizer, cctpL2toL1Finalizer],
   42161: [arbitrumOneFinalizer, cctpL1toL2Finalizer, cctpL2toL1Finalizer],
+  59144: [lineaL1ToL2Finalizer, lineaL2ToL1Finalizer],
```
Comment (nicholaspai) [inline src/finalizer/index.ts:70]: How does this change our environment variable configuration? Currently if I specify `FINALIZER_CHAINS=[1]` in my env I can finalize l1 to l2 linea. Does this mean I have to now specify `...=[59144]` to finalize linea?

Also, is there an easy way you think I can choose to ignore one of the two directions: l2-->l1 or l1 --> l2?

File: src/finalizer/utils/cctp/l1ToL2.ts (hunk 1 above)
Comment (james-a-morris) [reply src/finalizer/utils/cctp/l1ToL2.ts:68]: There's no use-case - I wouldn't push this through until we have the constants-v2 upgraded.

**Review by nicholaspai — APPROVED** [review]
I think this PR has tradeoffs. It makes configuration simpler but I lose ability to filter out certain finalizers. I guess I never had that choice for chains like 10 with multiple finalizers. So maybe its a non factor

File: src/finalizer/index.ts (hunk 2 above)
Comment (james-a-morris) [reply src/finalizer/index.ts:70]: We'd specify [59144] for L1 <-> L2 finalizations after this PR. If we maintained ```[1]: [lineaL1L2Finalizer, opCCTPL1L2Finalizer, ...]``` then enabling mainnet for finalization would mean we finalize all L1 -> L2 routes

File: src/finalizer/utils/cctp/l1ToL2.ts (hunk 1 above)
Comment (james-a-morris) [reply src/finalizer/utils/cctp/l1ToL2.ts]: https://github.com/across-protocol/relayer-v3/pull/1405/commits/2907cccb71cd4ec4e4c0c7f77bb2e06069bd4e7f

**Comment by james-a-morris** [comment]
Recent tests to confirm that the finalizer is in order post-upgrade of `constants-v2`:

1. https://polygonscan.com/tx/0x7a1f9993f1675c0a94fb194bec3929ed15f7d47d434b1a9741b022653c416423
2. https://basescan.org/tx/0x05b635c73c4e85b81e1bf8262fbdd632b367b0a3c1673db77aa887da057c9d01

**Review by nicholaspai — APPROVED** [review]

File: src/finalizer/index.ts [hunk 3]
```diff
@@ -123,20 +122,18 @@ export async function finalize(
...
         hubPoolClient.hubPool.address,
-        spokePoolClients[hubChainId === CHAIN_IDs.MAINNET ? CHAIN_IDs.LINEA : CHAIN_IDs.LINEA_GOERLI].spokePool.address,
         CONTRACT_ADDRESSES[hubChainId]?.atomicDepositor?.address,
-      ]);
+      ];
+      // For linea specifically, we want to include the Linea Spokepool as well.
+      if (sdkUtils.chainIsLinea(chainId)) {
+        addressesToEnsure.push(spokePoolClients[CHAIN_IDs.LINEA].spokePool.address);
+      }
+      l1ToL2AddressesToFinalize = enrichL1ToL2AddressesToFinalize(l1ToL2AddressesToFinalize, addressesToEnsure);
```
Comment (nicholaspai) [inline src/finalizer/index.ts:136]: For CCTP, do we need to track the spoke pools?

**Review by nicholaspai — COMMENTED** [review]
https://github.com/across-protocol/relayer-v3/pull/1405/files#r1572461159

File: src/finalizer/index.ts (hunk 3 above)
Comment (james-a-morris) [reply src/finalizer/index.ts:136]: We should only need to track the L1 wallets that are making deposits.

**Pull request merged**